modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C

```diff
@@ -0,0 +1,84 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+// StochasticTools includes
+#include "SamplerPostprocessorTransfer.h"
+#include "SamplerMultiApp.h"
+#include "SamplerReceiver.h"
+#include "StochasticResults.h"
+
+template <>
+InputParameters
+validParams<SamplerPostprocessorTransfer>()
+{
+  InputParameters params = validParams<MultiAppTransfer>();
+  params.addClassDescription("Transfers data to and from Postprocessors on the sub-application.");
+  params.addParam<VectorPostprocessorName>(
+      "results",
+      "Name of the StochasticResults object that the results are to be stored or extracted.");
+  params.addParam<std::string>(
+      "postprocessor", "Name of the Postprocessor on the sub-application to transfer to or from.");
+  params.set<MooseEnum>("direction") = "from_multiapp";
+  params.suppressParameter<MooseEnum>("direction");
+  return params;
+}
+
+SamplerPostprocessorTransfer::SamplerPostprocessorTransfer(const InputParameters & parameters)
+  : MultiAppTransfer(parameters),
+    _results_name(getParam<VectorPostprocessorName>("results")),
+    _sampler_multi_app(std::dynamic_pointer_cast<SamplerMultiApp>(_multi_app).get()),
+    _sampler(_sampler_multi_app->getSampler()),
+    _sub_pp_name(getParam<std::string>("postprocessor"))
+{
+  if (!_sampler_multi_app)
+    mooseError("The 'multi_app' must be a 'SamplerMultiApp.'");
+}
+
+void
+SamplerPostprocessorTransfer::initialSetup()
+{
+  const ExecuteMooseObjectWarehouse<UserObject> & user_objects = _fe_problem.getUserObjects();
+  std::shared_ptr<UserObject> uo = user_objects.getActiveObject(_results_name);
+  _results = std::dynamic_pointer_cast<StochasticResults>(uo).get();
+
+  if (!_results)
+    mooseError("The 'results' object must be a 'StochasticResults' object.");
+
+  _results->init(_sampler);
+}
+
+void
+SamplerPostprocessorTransfer::execute()
+{
+  // Number of PP is equal to the number of MultiApps
+  const unsigned int n = _multi_app->numGlobalApps();
+
+  // Collect the PP values for this processor
+  std::vector<PostprocessorValue> values;
+  values.reserve(_multi_app->numLocalApps());
+  for (unsigned int i = 0; i < n; i++)
+  {
+    if (_multi_app->hasLocalApp(i))
```

Comment:
This is OK, it certainly does fail. This is a consequence of how the MultiApps are stored and accessed. The Transfer API is based on global index, thus you must loop all the global index and only do work if you are local.

> If this fails, then the indexing between values and the multi_app numbering will be off.

They are not supposed to be in agreement, each processors is populating the values, then the values are assembled into one vector.

```
[ [N0_0, N0_1, ..., N0_n], [N1_0, N1,_1, .. N1_n], ...]
```

This does assume that MultiApps are distributed in a continuous fashion, which I believe is the case.